Filter out desynchronized packets

blm768 opened the pull request:
This is a work-in-progress attempt to filter out desynchronized packets instead of dropping the connection.

blm768 committed:
Filter out packets with invalid IDs

blm768 committed:
Prevent filtered-out packets from triggering empty commands

blm768 committed:
Actually shut down receiver loop

blm768 committed:
WIP: make sure connections are closed on shutdown

TODO: block in Drop impls if we haven't closed?

TODO: need to join background task

blm768 committed:
Don't put an error into the channel when nothing is listening

Avoids a case where the channel could fill up, blocking the receiving task.

blm768 committed:
Avoid deadlock when receiving unexpected packet type

blm768 committed:
Remove comment

ikkerens commented:
I gave the draft a look, and so far it looks good. As soon as you undraft I'll do some proper testing and whatnot. Thanks so far!

blm768 committed:
Remove Clone from ReconnectingConnection

This will help with later changes. Clone also doesn't really make sense as
part of the public API.

blm768 committed:
Join helper task when closing ReconnectingConnection

blm768 committed:
Simplify handling of unreachable branch

blm768 commented:
Still a good chance of logic bugs in the edge cases, but this at least seems to work on the "happy path".
